### PR TITLE
[MRG]Add UnifOrtho sampling of slicing directions for Sliced Wasserstein

### LINKED ISSUE #850: Add UnifOrtho sampling of slicing directions for Sliced Wasserstein
## 🚀 Feature

Add `UnifOrtho` sampling of slicing directions for Sliced Wasserstein --
recommended in the recent literature as the best choice in high
dimensions, complementing the QSW/RQSW (d=3-only) sampling added in #838.

### Motivation

QSW/RQSW (generalized spiral points, #838) only supports d=3. A recent,
dedicated variance analysis [2] gives an explicit recommendation for
higher dimensions:

> "Our final recommendation for the computation of the sliced
> Wasserstein distance is to use randomized quasi-Monte Carlo in low
> dimensions and UnifOrtho in large dimensions." -- Petrovic, Bardenet &
> Desolneux (2025)

`UnifOrtho` is the orthogonal Monte Carlo estimator introduced by
Rowland et al. (2019) [1]; [2] specifically analyses why it succeeds in
large dimensions and confirms it as the recommended choice there.

### Pitch

Add sampling from `UnifOrt(S^{d-1}; N)` [1]: N mutually orthogonal
directions, each marginally uniform on the sphere, obtained as N rows of
a Haar-random orthogonal matrix (concatenating independent draws when
N > d). No dimension restriction, unlike QSW/RQSW.

### Additional context

[1] Rowland, M., Hron, J., Tang, Y., Choromanski, K., Sarlos, T., &
    Weller, A. (2019). "Orthogonal Estimation of Wasserstein Distances."
    AISTATS 2019, PMLR 89.
[2] Petrovic, V., Bardenet, R., & Desolneux, A. (2025). "Repulsive
    Monte Carlo on the sphere for the sliced Wasserstein distance."
    arXiv:2509.10166 (under review at TMLR).

Follow-up to #838.

I'd like to work on this myself and will follow up with a PR.

### PR DESCRIPTION
**Pull request by Samuel-Vangu**
## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation (added or improved documentation for existing code)

## Motivation and context / Related issue

The Sliced Wasserstein distance only supports a low-discrepancy sampling
alternative to plain Monte Carlo (QSW/RQSW, PR #838) for `dim == 3`. This
leaves no principled alternative in higher dimension, where variance
reduction matters most.

Adds UnifOrtho sampling: independent blocks of mutually orthogonal
directions, drawn from the Haar measure on `O(dim)`, defined for any
`dim` and recommended specifically for large dimension by a recent
numerical and theoretical study. Reuses the existing `get_random_rotations`
directly, requiring no new randomization primitive.

Closes #850 .

## How has this been tested (if it applies)

- New unit tests covering: unit-norm output, exact mutual orthogonality
  within a block, correctness for arbitrary dimensions (including
  non-multiples of `dim`), seed reproducibility, cross-backend consistency
  (NumPy, PyTorch, JAX, TensorFlow), and reduced approximation error
  compared to uniform sampling in high dimension against a closed-form
  reference distance.
- Full existing test suite re-run to confirm no regression.
- New gallery example (`plot_uniortho.py`) demonstrating both the
  orthogonal block structure and the convergence gain in high dimension.

## PR checklist

- [x] I have read the [**CONTRIBUTING**](https://pythonot.github.io/contributing.html) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

**Comment by Samuel-Vangu** [comment]
Hi @rflamary @clbonet, 

this one's ready for review whenever you have time.
Not a big PR -- I've essentially reused existing code
(`get_random_rotations`) to add UnifOrtho sampling.

File: examples/sliced-wasserstein/plot_uniortho_high_dim.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+for ax in (ax1, ax2):
+    ax.set_box_aspect([1, 1, 1])
+    ax.view_init(elev=20, azim=45)
+    ax.set_xticks([])
+    ax.set_yticks([])
+    ax.set_zticks([])
+
+pl.tight_layout()
+pl.show()
```
Comment (clbonet) [inline examples/sliced-wasserstein/plot_uniortho_high_dim.py:124]: Add a sphere for better visualization?

File: examples/sliced-wasserstein/plot_uniortho_high_dim.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+UnifOrtho takes a different route that works in *any* dimension:
+instead of drawing directions independently, it draws them in blocks of ``dim`` directions,
+each block is a random orthonormal basis (a draw from the Haar measure on the
+orthogonal group :math:`O(\\mathrm{dim})`). Directions within a block are
+therefore exactly mutually orthogonal, spreading them out much more evenly
+than independent draws would.
+
+We first visualize this block structure on the ordinary 3D sphere, purely
+for intuition -- dimension 3 is precisely where QSW/RQSW should be
```
Comment (clbonet) [inline examples/sliced-wasserstein/plot_uniortho_high_dim.py:26]: You could maybe add this baseline in the 3D example to show that QMC is better than UnifOrtho in this case.

File: ot/sliced/_utils.py [hunk 3]
```diff
@@ -364,3 +364,91 @@ def projection_sphere_to_ball(x, eps=1e-6, backend=None):
     norm2 = nx.sum(x_azimuth**2, axis=-1, keepdims=True)
     radius = nx.arccos(-(3.0 + 5.0 * x_d) / (5.0 + 3.0 * x_d)) / np.pi
     return radius * x_azimuth / nx.sqrt(norm2)
+
+
+def get_projections_uniortho(d, n_projections, seed=None, backend=None, type_as=None):
```
Comment (clbonet) [inline ot/sliced/_utils.py]: get_random_orthogonal?

File: ot/sliced/_utils.py [hunk 4]
```diff
@@ -364,3 +364,91 @@ def projection_sphere_to_ball(x, eps=1e-6, backend=None):
...
+    # Number of independent orthogonal blocks needed so that k * d covers
+    # at least n_projections directions (integer ceiling of n_projections / d).
+    k = -(-n_projections // d)
+
+    # k independent Haar-uniform (d, d) orthogonal matrices. Within each
+    # (d, d) slice, the ROWS form d mutually orthogonal, unit-norm
+    # directions (see get_random_rotations for the QR-based construction
+    # and its Mezzadri (2007) sign correction, which is what makes the
+    # rotation genuinely Haar-uniform rather than merely orthogonal).
+    rotations = get_random_rotations(d, k, seed=seed, backend=nx, type_as=type_as)
```
Comment (clbonet) [inline ot/sliced/_utils.py:445]: This is on SO(d)?

File: ot/sliced/_sliced_distances.py [hunk 5]
```diff
@@ -118,6 +135,8 @@ def sliced_wasserstein_distance(
     .. [31] Bonneel, Nicolas, et al. "Sliced and radon wasserstein barycenters of measures." Journal of Mathematical Imaging and Vision 51.1 (2015): 22-45
     .. [95] Nguyen, K., Bariletto, N., & Ho, N. (2024). "Quasi-Monte Carlo for 3D Sliced Wasserstein." International Conference on Learning Representations (ICLR).
     .. [96] Rakhmanov, E. A., Saff, E. B., & Zhou, Y. M. (1994). "Minimal Discrete Energy on the Sphere." Mathematical Research Letters, 1(6), 647-662.
+    .. [97] Rowland, M., Hron, J., Tang, Y., Choromanski, K., Sarlos, T., & Weller, A. (2019). "Orthogonal Estimation of Wasserstein Distances." Proceedings of the 22nd International Conference on Artificial Intelligence and Statistics (AISTATS), PMLR 89.
+    .. [98] Petrovic, V., Bardenet, R., & Desolneux, A. (2025). "Repulsive Monte Carlo on the sphere for the sliced Wasserstein distance." arXiv:2509.10166.
```
Comment (clbonet) [inline ot/sliced/_sliced_distances.py]: This is a TMLR paper, add the right citation

Comment (clbonet) [reply ot/sliced/_sliced_distances.py]: I guess you can also cite https://openreview.net/forum?id=ECBepTWAFG which also studies methods to draw directions and tested the UnifOrth

**Review by clbonet — COMMENTED** [review]
Thank you @Samuel-Vangu for this PR. The code looks very good, I only have few minor comments.

**Comment by Samuel-Vangu** [comment]
Hi @clbonet, thanks for the thorough review! Here's what changed in
response:

- Renamed `get_projections_uniortho` to `get_random_orthogonal`
  everywhere (function, imports, tests, example, docstrings).
- Fixed the group: it's `SO(d)`, not `O(d)` . Corrected in the
  `get_random_orthogonal` and `sliced_wasserstein_distance` docstrings.
- Fixed the Petrovic et al. citation to point to TMLR instead of arXiv,
  and added the Sisouk, Delon & Tierny (2025) TMLR paper you linked as
  [99], wherever the text makes a general "the literature recommends"
  claim.
- Added a translucent sphere behind the scatter points in the 3D
  visualization.
- Added a compact d=3 convergence comparison (uniform vs UnifOrtho vs
  RQSW) right after the visualization, confirming numerically that RQSW
  wins there before moving on to the d=30 case where UnifOrtho takes
  over.

All 88 tests still pass.

File: ot/sliced/_utils.py [hunk 6]
```diff
@@ -364,3 +364,91 @@ def projection_sphere_to_ball(x, eps=1e-6, backend=None):
     norm2 = nx.sum(x_azimuth**2, axis=-1, keepdims=True)
     radius = nx.arccos(-(3.0 + 5.0 * x_d) / (5.0 + 3.0 * x_d)) / np.pi
     return radius * x_azimuth / nx.sqrt(norm2)
+
+
+def get_random_orthogonal(d, n_projections, seed=None, backend=None, type_as=None):
```
Comment (clbonet) [inline ot/sliced/_utils.py]: Sorry, I am hesitating, but I don't think it is clear with this name. It feels like we are sampling orthogonal matrices. Maybe use instead "get_random_orthogonal_directions"

File: README.md [hunk 7]
```diff
@@ -488,3 +488,8 @@ Artificial Intelligence.
 
 \[96] Rakhmanov, E. A., Saff, E. B., & Zhou, Y. M. (1994). [Minimal Discrete Energy on the Sphere](https://www.math.vanderbilt.edu/~esaff/texts/155.pdf). Mathematical Research Letters, 1(6), 647-662.
 
+\[97] Rowland, M., Hron, J., Tang, Y., Choromanski, K., Sarlos, T., & Weller, A. (2019). [Orthogonal Estimation of Wasserstein Distances](https://proceedings.mlr.press/v89/rowland19a.html). Proceedings of the 22nd International Conference on Artificial Intelligence and Statistics (AISTATS), PMLR 89:186-195.
+
+\[98] Petrovic, V., Bardenet, R., & Desolneux, A. (2025). [Repulsive Monte Carlo on the sphere for the sliced Wasserstein distance](https://arxiv.org/abs/2509.10166). Transactions on Machine Learning Research.
```
Comment (clbonet) [inline README.md]: Use the link to the TMLR paper https://openreview.net/forum?id=JSiTmB6Ehu and I think the year is 2026.

**Review by clbonet — COMMENTED** [review]
Thank you @Samuel-Vangu for the update.

Sorry, I have one last comment. I think `get_random_orthogonal` is a bit misleading, I would suggest `get_random_orthogonal_directions` instead?

**Comment by Samuel-Vangu** [comment]
@clbonet 

Renamed `get_random_orthogonal` to `get_random_orthogonal_directions`
everywhere (function, imports, tests, example, docstrings). Fixed the
Petrovic et al. citation to point to the OpenReview page
(https://openreview.net/forum?id=JSiTmB6Ehu) and corrected the year to
2026 -- in `_utils.py`, `_sliced_distances.py`, the example, and
README.md.

All 88 tests still pass.

**Review by clbonet — APPROVED** [review]
Thank you @Samuel-Vangu, it looks good to me!
